### ISSUE TITLE
strictPropertyInitialization: false positives when using separate if clause for check

### ISSUE THREAD

**Issue by AlCalzone** [Microsoft/TypeScript#30992]
**TypeScript Version:**  3.4.2

**Search Terms:** strictPropertyInitialization, throw, false positive

**Code**
Consider all of the following examples
```ts
class Foo1 {
  constructor(arg: number | string | undefined) {
    if (arg == undefined) {
      throw new Error();
    }
    if (typeof arg === "number") {
      this.version = 1;
    } else if (typeof arg === "string") {
      this.version = 2;
    } else {
      arg; // inferred as never
    }
  }

  public version: number;
}

class Foo2 {
  constructor(arg: number | string | undefined) {
    if (arg == undefined) {
      arg = Math.random() > 0.5 ? 1 : undefined;
      if (arg == undefined) throw new Error();
    }
    if (typeof arg === "number") {
      this.version = 1;
    } else if (typeof arg === "string") {
      this.version = 2;
    } else {
      arg; // inferred as never
    }
  }

  public version: number;
}


class Foo3 {
  constructor(arg: number | string | undefined) {
    if (arg == undefined) {
      arg = 1;
    }
    if (typeof arg === "number") {
      this.version = 1;
    } else if (typeof arg === "string") {
      this.version = 2;
    } else {
      arg; // inferred as never
    }
  }

  public version: number;
}
```

**Expected behavior:**
In all 3 cases, `version` is definitely assigned in the constructor or the constructor throws.

**Actual behavior:**
TypeScript reports `Property 'version' has no initializer and is not definitely assigned in the constructor.`

**Playground Link:** 
turn on `strictPropertyInitialization` and `strictNullChecks`
https://www.typescriptlang.org/play/#src=class%20Foo1%20%7B%0D%0A%20%20constructor(arg%3A%20number%20%7C%20string%20%7C%20undefined)%20%7B%0D%0A%20%20%20%20if%20(arg%20%3D%3D%20undefined)%20%7B%0D%0A%20%20%20%20%20%20throw%20new%20Error()%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20if%20(typeof%20arg%20%3D%3D%3D%20%22number%22)%20%7B%0D%0A%20%20%20%20%20%20this.version%20%3D%201%3B%0D%0A%20%20%20%20%7D%20else%20if%20(typeof%20arg%20%3D%3D%3D%20%22string%22)%20%7B%0D%0A%20%20%20%20%20%20this.version%20%3D%202%3B%0D%0A%20%20%20%20%7D%20else%20%7B%0D%0A%20%20%20%20%20%20arg%3B%20%2F%2F%20inferred%20as%20never%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%0D%0A%0D%0A%20%20public%20version%3A%20number%3B%0D%0A%7D%0D%0A%0D%0Aclass%20Foo2%20%7B%0D%0A%20%20constructor(arg%3A%20number%20%7C%20string%20%7C%20undefined)%20%7B%0D%0A%20%20%20%20if%20(arg%20%3D%3D%20undefined)%20%7B%0D%0A%20%20%20%20%20%20arg%20%3D%20Math.random()%20%3E%200.5%20%3F%201%20%3A%20undefined%3B%0D%0A%20%20%20%20%20%20if%20(arg%20%3D%3D%20undefined)%20throw%20new%20Error()%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20if%20(typeof%20arg%20%3D%3D%3D%20%22number%22)%20%7B%0D%0A%20%20%20%20%20%20this.version%20%3D%201%3B%0D%0A%20%20%20%20%7D%20else%20if%20(typeof%20arg%20%3D%3D%3D%20%22string%22)%20%7B%0D%0A%20%20%20%20%20%20this.version%20%3D%202%3B%0D%0A%20%20%20%20%7D%20else%20%7B%0D%0A%20%20%20%20%20%20arg%3B%20%2F%2F%20inferred%20as%20never%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%0D%0A%0D%0A%20%20public%20version%3A%20number%3B%0D%0A%7D%0D%0A%0D%0A%0D%0Aclass%20Foo3%20%7B%0D%0A%20%20constructor(arg%3A%20number%20%7C%20string%20%7C%20undefined)%20%7B%0D%0A%20%20%20%20if%20(arg%20%3D%3D%20undefined)%20%7B%0D%0A%20%20%20%20%20%20arg%20%3D%201%3B%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20if%20(typeof%20arg%20%3D%3D%3D%20%22number%22)%20%7B%0D%0A%20%20%20%20%20%20this.version%20%3D%201%3B%0D%0A%20%20%20%20%7D%20else%20if%20(typeof%20arg%20%3D%3D%3D%20%22string%22)%20%7B%0D%0A%20%20%20%20%20%20this.version%20%3D%202%3B%0D%0A%20%20%20%20%7D%20else%20%7B%0D%0A%20%20%20%20%20%20arg%3B%20%2F%2F%20inferred%20as%20never%0D%0A%20%20%20%20%7D%0D%0A%20%20%7D%0D%0A%0D%0A%20%20public%20version%3A%20number%3B%0D%0A%7D%0D%0A

**Related Issues:**

**Comment by j-oliveras** [comment]
The `never` branches don't assign it. If you make that branches throw it works.

**Comment by ahejlsberg** [comment]
This is working as intended. Our reachability analysis is based on the _grammatical_ structure of the code. So, if you add a `throw` we know for certain that any following code isn't reachable. One way to think of it is that since there is no run-time validation of types, it is indeed possible for control to reach the final `else` clause. However, from a _static_ analysis perspective, the types don't permit it and therefore the type is `never` in that branch.

**Comment by AlCalzone** [comment]
Thanks for the explanation. I keep forgetting that control flow analysis is (somewhat) accounting for wrong runtime types.
